File: prog2/bhax/thematic_tutorials/bhax_textbook/src/liskov/downcasting.java
class Base {
    public void basePrint() {
        System.out.println("base class");
    }
}
class Derived extends Base {
    public void  derivedPrint() {
        System.out.println("derived class");
    }
}

public class downcasting {

    public static void main(String[] args) {
        Derived dClass = new Derived();
        dClass.basePrint();
        dClass.derivedPrint();
        ((Base)dClass).basePrint();
        //((Base)dClass).derivedPrint(); //downcast esetén itt hibát kapunk
    }
}

File: prog2/bhax/thematic_tutorials/bhax_textbook/src/liskov/downcasting.cpp
#include <iostream>

class Base
{
public:
	void baseMethod()
	{
		std::cout << "Base funcion\n";
	}
};

class Derived : public Base
{
public:
	void derivedMethod()
	{
		std::cout << "Derived funcion\n";
	}
};

int main()
{
	Derived test;
	test.baseMethod();
	test.derivedMethod();
	Base* test1 = (Base*)&test;
	test1->baseMethod();
	//test1->derivedMethod(); //hibát okoz

	return 0;
}
